File: ios/Sources/CompanionCore/ActivityRuns.swift
import Foundation

/// A grouped item in a chat transcript — either a single message or a folded run of tool activities.
public enum TranscriptItem: Identifiable, Hashable, Sendable {
    case message(Message)
    case run(id: String, messages: [Message])

    public var id: String {
        switch self {
        case let .message(msg): return msg.id
        case let .run(id, _): return id
        }
    }

    public var messages: [Message] {
        switch self {
        case let .message(msg): return [msg]
        case let .run(_, msgs): return msgs
        }
    }

    public var date: Date {
        switch self {
        case let .message(msg): return msg.date
        case let .run(_, msgs): return msgs.first?.date ?? Date()
        }
    }
}

/// A step that may be folded into an activity run.
public func isFoldableActivity(_ message: Message) -> Bool {
    guard message.kind == .activity, let tool = message.tool else { return false }
    if message.comm != nil { return false }
    return !tool.name.hasPrefix("error:")
}

/// Fold consecutive tool activities into grouped runs.
public func groupActivityRuns(_ messages: [Message]) -> [TranscriptItem] {
    var items: [TranscriptItem] = []
    var run: [Message] = []

    func flush() {
        if run.count > 1 {
            items.append(.run(id: "run:\(run[0].id)", messages: run))
        } else {
            for msg in run {
                items.append(.message(msg))
            }
        }
        run.removeAll()
    }

    for message in messages {
        if isFoldableActivity(message) {
            if let first = run.first {
                if first.role != message.role ||
                    first.from?.botId != message.from?.botId {
                    flush()
                }
            }
            run.append(message)
            continue
        }
        flush()
        items.append(.message(message))
    }
    flush()
    return items
}

/// Describes a folded activity run with tool breakdown and failure count.
public func describeActivityRun(_ messages: [Message]) -> (headline: String, summary: String, failedCount: Int) {
    var counts: [(name: String, count: Int)] = []
    for msg in messages {
        let name = msg.tool?.name ?? "step"
        if let idx = counts.firstIndex(where: { $0.name == name }) {
            counts[idx].count += 1
        } else {
            counts.append((name: name, count: 1))
        }
    }
    let parts = counts.map { $0.count > 1 ? "\($0.name) ×\($0.count)" : $0.name }
    let maxShown = 3
    let shown = parts.prefix(maxShown).joined(separator: ", ")
    let rest = parts.count > maxShown ? " +\(parts.count - maxShown) more" : ""
    let failed = messages.filter { $0.tool?.ok == false }.count

    let headline = "\(messages.count) \(messages.count == 1 ? "tool call" : "tool calls")"
    let summary = "\(shown)\(rest)"
    return (headline: headline, summary: summary, failedCount: failed)
}

File: src/lib/activity-runs.ts
// Grouping a transcript's tool chips into runs.
//
// A bot working through a task emits one chip per tool call, and a long
// stretch of them buries the thing you actually came to read: what the bot
// SAID. Consecutive finished steps fold into a single row that names them;
// text between two stretches breaks the run, so the bot's words always
// separate one run from the next.
import type { Message } from "@/state/store";
import { classifyTool, toolVerb } from "../../shared/tool-activity";

export type TranscriptItem =
  | { kind: "message"; message: Message }
  | { kind: "run"; id: string; messages: Message[] };

/** A step that may be folded away: finished, a real tool, and not a
 * bot⇄bot chip (those are navigation, not work) or a failed turn (that
 * renders as an error). A step still running stays out, so live progress
 * is never hidden behind a fold. */
function foldable(message: Message): boolean {
  const tool = message.tool;
  if (message.kind !== "activity" || !tool) return false;
  if (message.comm) return false;
  if (tool.ok !== true) return false;
  return !tool.name.startsWith("error:");
}

/** Runs shorter than this stay unfolded.
 *
 * A step is one line now, so five of them cost less vertical space than the
 * fold that would hide them — and a reader who can see "Read · config.ts,
 * Read · store.ts" does not have to click to learn the turn was working
 * through the codebase.  Folding earns its place only when a run is long
 * enough to bury the bot's words. */
export const RUN_FOLD_MIN = 6;

export function groupActivityRuns(messages: Message[]): TranscriptItem[] {
  const items: TranscriptItem[] = [];
  let run: Message[] = [];
  const flush = () => {
    // a short run is cheaper to read than a fold that hides it
    if (run.length >= RUN_FOLD_MIN) items.push({ kind: "run", id: `run:${run[0].id}`, messages: run });
    else for (const message of run) items.push({ kind: "message", message });
    run = [];
  };
  for (const message of messages) {
    if (foldable(message)) {
      const first = run[0];
      if (
        first &&
        (first.role !== message.role ||
          first.from?.botId !== message.from?.botId ||
          new Date(first.at).toDateString() !== new Date(message.at).toDateString())
      ) {
        flush();
      }
      run.push(message);
      continue;
    }
    flush();
    items.push({ kind: "message", message });
  }
  flush();
  return items;
}

const MAX_NAMES = 3;
const MAX_TARGETS = 2;

/** The one line a folded run has to earn its place with: how much work it
 * was, which tools did it, and whether anything failed — the last being the
 * only reason you would open it. */
export function describeRun(messages: Message[]): string {
  const counts = new Map<string, number>();
  for (const message of messages) {
    const tool = message.tool;
    // classify here too, so a fold header names the work the same way the
    // rows inside it do — including for steps recorded before the harness
    // started reporting a kind
    const label = tool ? toolVerb(tool.kind ?? classifyTool(tool.name), tool.name) : "";
    counts.set(label, (counts.get(label) ?? 0) + 1);
  }
  const names = [...counts].map(([name, count]) => (count > 1 ? `${name} ×${count}` : name));
  const shown = names.slice(0, MAX_NAMES).join(", ");
  const rest = names.length > MAX_NAMES ? ` +${names.length - MAX_NAMES} more` : "";
  const failed = messages.filter((message) => message.tool?.ok === false).length;
  // naming what it touched is the difference between "7 steps" and "7 steps
  // through the driver files" — the second one a reader can skip on purpose
  const targets = [
    ...new Set(messages.flatMap((message) => (message.tool?.target ? [message.tool.target] : []))),
  ];
  const where = targets.length
    ? ` — ${targets.slice(0, MAX_TARGETS).map(basename).join(", ")}${
        targets.length > MAX_TARGETS ? ` +${targets.length - MAX_TARGETS}` : ""
      }`
    : "";
  return `${messages.length} steps · ${shown}${rest}${where}${failed ? ` · ${failed} failed` : ""}`;
}

/** Last path segment, for a fold header that has room for two of them. */
function basename(value: string): string {
  const path = value.split(/[?#]/)[0] ?? value;
  const parts = path.split(/[/\\]/).filter(Boolean);
  const last = parts.at(-1) ?? value;
  return last.length > 28 ? last.slice(0, 27) + "…" : last;
}
